libasherah_test.go

```diff
@@ -0,0 +1,680 @@
+package main
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/godaddy/cobhan-go"
+)
+
+func validSetupForTesting(t *testing.T) {
+	config := &Options{}
+
+	config.KMS = "static"
+	config.ServiceName = "TestService"
+	config.ProductID = "TestProduct"
+	config.Metastore = "memory"
+	config.EnableSessionCaching = true
+	config.SessionCacheDuration = 1000
+	config.SessionCacheMaxSize = 2
+	config.ExpireAfter = 1000
+	config.CheckInterval = 1000
+	config.Verbose = true
+	config.RegionMap = RegionMap{}
+	config.RegionMap.UnmarshalFlag("region1=arn1,region2=arn2")
+
+	buf := cobhan.AllocateBuffer(4096)
+	result := cobhan.JsonToBufferSafe(config, &buf)
+	if result != ERR_NONE {
+		t.Errorf("BytesToBufferSafe returned %v", result)
+	}
+
+	SetupJson(cobhan.Ptr(&buf))
+}
+
+func testAllocateStringBuffer(t *testing.T, str string) []byte {
+	buf, result := cobhan.AllocateStringBuffer(str)
+	if result != ERR_NONE {
+		t.Error(fmt.Sprintf("AllocateStringBuffer returned %v", result))
+		t.FailNow()
+	}
+	return buf
+}
+
+func TestSetupJson(t *testing.T) {
+	validSetupForTesting(t)
+	Shutdown()
+}
+
+func TestSetupJson2(t *testing.T) {
+	config := &Options{}
+
+	config.KMS = "static"
+	config.ServiceName = "TestService"
+	config.ProductID = "TestProduct"
+	config.Metastore = "memory"
+	config.EnableSessionCaching = true
+	config.Verbose = false
+
+	buf := cobhan.AllocateBuffer(4096)
+	result := cobhan.JsonToBufferSafe(config, &buf)
+	if result != ERR_NONE {
+		t.Errorf("BytesToBufferSafe returned %v", result)
+	}
+
+	SetupJson(cobhan.Ptr(&buf))
+	Shutdown()
+}
+
+func TestSetupJsonTwice(t *testing.T) {
+	config := &Options{}
+
+	config.KMS = "static"
+	config.ServiceName = "TestService"
+	config.ProductID = "TestProduct"
+	config.Metastore = "memory"
+	config.EnableSessionCaching = true
+	config.Verbose = true
+
+	buf := cobhan.AllocateBuffer(4096)
+	result := cobhan.JsonToBufferSafe(config, &buf)
+	if result != ERR_NONE {
+		t.Errorf("BytesToBufferSafe returned %v", result)
+	}
+
+	SetupJson(cobhan.Ptr(&buf))
+	SetupJson(cobhan.Ptr(&buf))
```

Comment:
Shouldn't this second call be erroring with `ERR_ALREADY_INITIALIZED`?

Reply:
Fixed thanks